Add support for chart filtering via `metrics` query param

robhudson opened the pull request:
This isn't quite ready yet but I'm using it as a way to lint check and run the tests while developing.

robhudson commented:
I consider this ready for review now. @openjck wanna take it? If there are any improvements to design/style I'd prefer to merge then let one of you from the front-end make things nicer. Functionally this works as I had hoped.

openjck commented:
Awesome! I'll take a look.

openjck commented:
Major kudos for taking up a feature this big with a new technology.

openjck commented on distributionviewer/core/static/js/app/api/metric-api.js:
```diff
@@ -14,15 +14,39 @@ const mockEndpoints = {
+// Get query param metrics as CSV of ints
+export function getQueryMetrics(query) {
+  var qmetrics = [];
+  if (query !== undefined && query.metrics !== undefined) {
```
Alternatively `if (query && query.metrics)`

openjck commented on distributionviewer/core/static/js/app/api/metric-api.js:
```diff
@@ -14,15 +14,39 @@ const mockEndpoints = {
+  if (query !== undefined && query.metrics !== undefined) {
+    // We were given a list of metrics to display, filter by those.
+    qmetrics = query.metrics.split(',').map(
+      function(v) {
```
Just a nitpick. ES6 gives us a new syntax for lambdas, which we have the option of using here to make the code a little more concise. It would look something like this:

``` javascript
qmetrics = query.metrics.split(',').map(v => parseInt(v, 10));
```

openjck commented on distributionviewer/core/static/js/app/api/metric-api.js:
```diff
@@ -14,15 +14,39 @@ const mockEndpoints = {
 
 export const endpoints = process.env.NODE_ENV === 'production' ? prodEndpoints : mockEndpoints;
 
+// Get query param metrics as CSV of ints
```
Just a nitpick. It looks like this takes a CSV of `int`s and returns an array. I was a bit confused by the comment.

openjck commented on distributionviewer/core/static/js/app/api/metric-api.js:
```diff
@@ -14,15 +14,39 @@ const mockEndpoints = {
+    var qmetrics = getQueryMetrics(query);
+    var metrics = response.data.metrics;
+
+    if (qmetrics.length > 0) {
```
Alternatively `if (qmetrics.length)`

openjck commented on distributionviewer/core/static/js/app/api/metric-api.js:
```diff
@@ -14,15 +14,39 @@ const mockEndpoints = {
+          return qmetrics.indexOf(metric.id) > -1;
+        }
+      );
+    }
```
We also have the option of using an arrow function here:

``` javascript
metrics = metrics.filter(metric => qmetrics.indexOf(metric.id) > -1));
```

openjck commented on distributionviewer/core/static/js/app/api/metric-api.js:
```diff
@@ -14,15 +14,39 @@ const mockEndpoints = {
   return axios.get(endpoints.GET_METRICS).then(response => {
-    store.dispatch(metricActions.getMetricsMetadataSuccess(response.data.metrics));
+    var qmetrics = getQueryMetrics(query);
+    var metrics = response.data.metrics;
```
Just a nitpick. We could consider calling this `metricsMetadata` since we ultimately dispatch it using `getMetricsMetadataSuccess`. By convention, we've been using the term _metadata_ to refer to the output of `/metrics/` and the term _data_ to refer to the output of `/metrics/#/`. Maybe there's a totally different naming scheme that would be even better.

openjck commented on distributionviewer/core/static/js/app/components/containers/chart-config-container.js:
```diff
@@ -0,0 +1,38 @@
+      input => input.value
+    ).join(',');
+
+    if (metrics !== '') {
```
Alternatively `if (metrics)`

openjck commented on distributionviewer/core/static/js/app/components/containers/chart-config-container.js:
```diff
@@ -0,0 +1,38 @@
+    metricApi.getMetrics();
+  }
+
+  handleSubmit(e) {
```
By convention private methods start with an underscore, like `_handleSubmit()`. I'll admit I haven't been consistent about this myself.

openjck commented on distributionviewer/core/static/js/app/components/layouts/main-layout.js:
```diff
@@ -9,7 +9,10 @@ export default function(props) {
+        <h1>
+          <Link className="primary-header-link" to="/">Firefox Distribution Viewer</Link>
+          <Link to={`/configure/?metrics=${props.location.query.metrics}`}><img className="gear" width="30" height="30" title="Dashboard Configuration" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAB4AAAAeCAYAAAA7MK6iAAAABGdBTUEAALGPC/xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAACXBIWXMAAA7EAAAOxAGVKw4bAAABWWlUWHRYTUw6Y29tLmFkb2JlLnhtcAAAAAAAPHg6eG1wbWV0YSB4bWxuczp4PSJh ...
+        </h1>
```
Semantically, we could consider moving the second link and the gear image out of the `h1` since only the title really represents top-level heading content. Per MDN, "a heading element briefly describes the topic of the section it introduces." It might be a little more difficult to style, though.

In fact, `<Link className="primary-header-link" to="/"><h1>Firefox Distribution Viewer</h1></Link>` is probably the most semantically correct but it's not technically valid.

robhudson replied:
Is this another area where we can clean it up after merge? I think it might actually be better placed near the sign out at top/right but when I tried to put it there all the CSS broke and I decided that this was good enough to show the feature working for now. :)

openjck replied:
Yeah, no worries. I can follow up with that. I'm being pretty pedantic anyway. Don't fret over it.

openjck commented on distributionviewer/core/static/js/app/components/views/chart-config.js:
```diff
@@ -0,0 +1,18 @@
+      {props.metadata.map(metric => {
+        var checked = props.qmetrics.indexOf(metric.id) > -1;
+        return (
+          <label key={metric.id}><input type="checkbox" defaultChecked={checked} name="metrics" value={metric.id} /> {metric.name}: {metric.description}</label>
```
We could remove the extra space after `<input>`

openjck reviewed (changes requested):
A lot of comments, but don't get disheartened. Most are nitpicks. This is actually really well done, especially considering this is your first major feature with React!

openjck commented on distributionviewer/core/static/js/app/components/containers/chart-config-container.js:
```diff
@@ -0,0 +1,38 @@
+
+class ChartConfigContainer extends React.Component {
+  componentDidMount() {
+    metricApi.getMetrics();
```
This isn't your fault, but the way I wrote `getMetrics`, it dispatches both the metadata (the output of `/metrics/`) and the data (the output of each `/metric/#/`). We only need the metadata here, so we could refactor the API to allow for that. I'll follow up with a PR after this lands.

openjck commented on distributionviewer/core/static/js/app/components/containers/chart-config-container.js:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  render() {
+    var qmetrics = metricApi.getQueryMetrics(this.props.location.query);
```
As far as I can tell (I might be missing something), we could put this into `componentDidMount` and it would still work. It would look something like this:

``` javascript
componentDidMount() {
  metricApi.getMetrics();
  this.qmetrics = metricApi.getQueryMetrics(this.props.location.query);
}

// ...

render() {
  return (
    <ChartConfig {...this.props} qmetrics={this.qmetrics} handleSubmit={this.handleSubmit} />
  );
}
```

A small change, but it would allow the component to re-render if it needs to for any reason without calling `getQueryMetrics` again.

robhudson replied:
This gave me an error but I put it in the `constructor` and it worked nicely.

robhudson commented:
Awesome comments, thanks! I'll reply to a few and start fixing the rest.

spasovski commented on distributionviewer/core/static/css/chart-config.styl:
```diff
@@ -0,0 +1,22 @@
+@import 'lib';
+
+#chart-config {
```
IDs are a bit too specific which makes overriding them tricky later. The rest seems good though.

Commit:
Add dashboard configuration (Fixes #1)

This adds support for filtering the list of charts via the `metrics`
query parameter, along with a configuration view to select which charts
to display.

robhudson commented:
Updated

openjck commented:
This is so good. Super useful and works flawlessly. Really nice work.
